d4r for design automation

pinzart90 opened the pull request:
Minimal changes to get D4R working on Design automation
This PR is based on RC3.3.0_Revit2025 because the latest Revit on DA seems to be 2025

I opted to use "if def"s in order to minimize the changes and avoid breaking changes. As a follow up task we should do a proper separation of Core and UI functionality in all the projects (this will most likely mean breaking changes).
I will create a couple of jira tasks with proposed changes to pursue this further.

zeusongit committed:
updated GregRevitAuth (#3027)

vpavel04 committed:
fix two failed tests

Mikhinja committed:
bumpup version to make build with signed binaries

Mikhinja committed:
bump up version to publish

aparajit-pratap committed:
Update DynamoRevit.cs (#3033)

OnRequestsCrashPrompt function

Mikhinja committed:
Update Dynamo Core to 3.0.2.7476

Mikhinja committed:
Update DynamoRevit versioning to 3.0.2 to align with DynamoCore

Mikhinja committed:
Fix a couple of Regression tests

Mikhinja committed:
Update Dynamo Core to 3.0.3.7597

Mikhinja committed:
Update DynamoRevit versioning to 3.0.3

Mikhinja committed:
Version *.customization.dll files with number from other output binary

saintentropy committed:
Fix regression in familytype.bygeometry (#3066)

Mikhinja committed:
Integrate Dynamo Core 3.1 in Revit 2025.2

Mikhinja committed:
Update D4R minor version to 3.1.0

Mikhinja committed:
Update Dynamo to 3.2.0
Update Greg to 3.0.1

Mikhinja committed:
Update Dynamo to 3.2.0
Update Greg to 3.0.1

Mikhinja committed:
Update Dynamo to 3.2.0

Mikhinja committed:
* update Dynamo Core to 3.2.1
* update GregRevitAuth to 3.0.8935.26399

Mikhinja committed:
* update Dynamo Core to 3.2.1
* update GregRevitAuth to 3.0.8935.26399

Mikhinja committed:
Merge branch 'RC3.2.0_Revit2025' of https://github.com/DynamoDS/DynamoRevit into RC3.2.0_Revit2025

Mikhinja committed:
REVIT-222587
* add OOTB templates
* update Dynamo Core build to 3.2.1.5366

Mikhinja committed:
change content publishing

AdrianaStanica committed:
Roof.Points bug fix integration in Revit 2025.3 (#3081)

* add wrapper over Revit RoofBase class to D4R Roof class

* cast value data type to parameter value data type in order to avoid blind pointer cast in revit later on

* cast value data type to parameter value data type, get parameter value data type from Storage type where value is null

Mikhinja committed:
update Dynamo Core to 3.2.2.5494 (#3084)

ZbynekHanak-RWS committed:
Create ContinuousLocalization.yml

Mikhinja committed:
Update Dynamo Core to 3.3.0

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
Update RevitNodesUI.csproj

pinzart90 committed:
UPDATE

mjkkirschner commented on src/DADynamoConsole/Program.cs:
```diff
@@ -0,0 +1,4 @@
+﻿using System.Reflection;
+
+Assembly.LoadFrom(@"C:\Program Files\Autodesk\Revit 2025\RevitAPI.dll");
```
are these console applications / entry points used? How does DA call them? Won't these not run the in context of Revit but as separate processes?

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 commented on src/Libraries/RevitNodes/RevitNodeModel.cs:
```diff
@@ -11,4 +11,4 @@ public RevitNodeModel() { }
         [JsonConstructor]
         public RevitNodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts) { }
     }
```
@mjkkirschner do you think this a proper place to have a NodeModel base class ?

pinzart90 replied:
It is used in RevitNodesUI (valid for cloud context too)

pinzart90 committed:
update

pinzart90 committed:
Update DisposeLogic.cs

pinzart90 committed:
update

pinzart90 committed:
update

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+        private string WorkItemFolder;
+
+        private static EventHandlerProxy proxy;
+        public static EventHandlerProxy EventHandlerProxy
```
/// summary

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            get { return proxy; }
+        }
+
+        public static List<IUpdater> Updaters = new List<IUpdater>();
```
why is this static, public etc?

pinzart90 replied:
no reason, changing

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+
+        private void CheckIfPythonExists(string inputFolder)
+        {
+            var pyDir = "python-3.9.12-embed-amd64";
```
maybe this should be a bit more tolerant of different python versions?

pinzart90 replied:
I am not checking this anymore

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+
+        public ExternalDBApplicationResult OnStartup(ControlledApplication application)
+        {
+            AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
```
Maybe we should just stop using AppDomain - it's just a pointer to the default ALC - right? Maybe lets be more specific, especially if Revit starts using different ALCs for Dynamo/Addins?

pinzart90 replied:
As far as i know, AppDomain.CurrentDomain is the only API that can add an `UnhandledException ` handler

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            try
+            {
+                RevitServices.Transactions.TransactionManager.SetupManager(new RevitServices.Transactions.AutomaticTransactionStrategy());
+                ElementBinder.IsEnabled = true;
```
I am not sure element binding really makes sense in the context of design automation, it really only makes sense in interactive workflows.

I would disable it by default or potentially have a config option to turn it on instead of defaulting to on.

pinzart90 replied:
I guess, if the workers are stateless and thorw away.

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+                RevitServices.Transactions.TransactionManager.SetupManager(new RevitServices.Transactions.AutomaticTransactionStrategy());
+                ElementBinder.IsEnabled = true;
+
+                RevitServicesUpdater.Initialize(Updaters);
```
IMO updaters seem similar to element binding, the use case I am familiar with for `updaters` is that when updaters raise changes Dynamo will kick off node executions, but this doesn't make much sense in a non-interactive single run session.

Let's chat about this more - we should review how updaters are used in DynamoRevit and what use cases we need to support in Design Automation.

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+
+                RevitServicesUpdater.Initialize(Updaters);
+
+                DesignAutomationBridge.DesignAutomationReadyEvent += HandleDesignAutomationReadyEvent;
```
I guess unsubscribing this doesn't really matter since DA will shutdown Revit between runs?

pinzart90 replied:
that was my understanding, but does not hurt to cleanup

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            Console.WriteLine("<<!>> Preparing Dynamo model. Vers 1");
+
+            var hostloc = typeof(Autodesk.Revit.ApplicationServices.Application).Assembly.Location;
+            var asmLocation = Path.GetDirectoryName(hostloc);
```
this isn't going to work with ASC (autodesk shared components)
@sm6srw can you confirm?

sm6srw replied:
This will not work on desktop Revit. But I am not sure if they use ASC in design automation. We don't do the ASC lookup in D4R normally. Revit tells us where it is. Does that not work in this context?

pinzart90 replied:
Is there a Revit API that gives us the ASC location?

Mikhinja replied:
Yes, `Autodesk.Revit.ApplicationServices.ControlledApplication.SharedComponentsLocation`.

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            var asmLocation = Path.GetDirectoryName(hostloc);
+            Console.WriteLine($"using asm at location {asmLocation}");
+            Console.WriteLine($"Is Loaded {LoadMessage}");
+            var curFolder = Directory.GetParent(Assembly.GetExecutingAssembly().Location).FullName;
```
I'm confused what directory this is supposed to be pointing to - current in what sense?

pinzart90 replied:
Changed most of this, I think I was refering to the current working directory

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            cmdArgs.DisableAnalytics = true;
+            cmdArgs.ServiceMode = true;
+            cmdArgs.AnalyticsInfo = new() { HostName = "Revit_DA" };
+            cmdArgs.ImportedPaths = [Path.Combine(curFolder, "nodes", "RevitNodes.dll"), Path.Combine(curFolder, "nodes", "DSRevitNodesUI.dll")];
```
okay, you're going to need to give me some more info as to why you think this approach is better than what Dynamo/DynamoRevit does normally to resolve dependencies?

Or is the crux, that you want to avoid subclassing DynamoModel for some reason?

pinzart90 replied:
reverted most of this

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            cmdArgs.ServiceMode = true;
+            cmdArgs.AnalyticsInfo = new() { HostName = "Revit_DA" };
+            cmdArgs.ImportedPaths = [Path.Combine(curFolder, "nodes", "RevitNodes.dll"), Path.Combine(curFolder, "nodes", "DSRevitNodesUI.dll")];
+            // need this for cloud, does not work on local
```
what does not work in local?

pinzart90 replied:
I could not figure out how to make the dynamo userDataFolder and commonDataFolder to work with the DA work directory both in cloud and locally (I probably just lost my patience)

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+            cmdArgs.CommonDataFolder = Path.Combine(dynTempDir, "Dynamo");
+
+            // Startup a new project, maybe an option we can have ?
+            //app.NewProjectDocument(Autodesk.Revit.DB.UnitSystem.Metric);
```
If I were a user executing against dynamo as a service - I would expect this information to somehow have been supplied via a web UI or via part of the graph authoring experience ahead of time.

For example:

1. I authored the graph in a metric project, so that should be taken into account when run in this context against another project.
2. I set this graph to run against a particular project already so we should just open that project.


Do you have a user use case in mind where a graph should be run on a brand new project?

pinzart90 replied:
I don't have a use case for NewProjectDocument yet. 
Maybe if the user wants to generate data from scratch somehow.

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+
+            LoadMessage = model != null ? "loaded" : "no loaded";
+            Console.WriteLine($"Checking python in {curFolder}");
+            CheckIfPythonExists(curFolder);
```
can you explain why you are doing this as opposed to letting Dynamo load python the normal way?

pinzart90 replied:
I changed some of this, but I am loading python from a subfolder (wherever the DA engine unzips the pythonLibs zip)
Dynamo has an issue where it expects all python libs should be in the same folder as DynamoCore.dll (else it throws an error)

mjkkirschner commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+
+            File.WriteAllText(Path.Combine(WorkItemFolder, "result.json"), result);
+
+            model.RunCompleted += Model_RunCompleted;
```
isn't this too late to attach this event? Hasn't the graph already been run via HandleRoute?

pinzart90 replied:
maybe, but I'm not doing much with it 
Will remove it for now

mjkkirschner reviewed (commented):
part 1 of my review - mostly questions. Will continue pt2 when I can.

mjkkirschner commented on src/DADynamoApp/DADynamoApp.csproj:
```diff
@@ -0,0 +1,47 @@
+    </ItemGroup>
+    <ItemGroup>
+        <PackageReference Include="Autodesk.Forge.DesignAutomation.Revit" Version="2025.0.1" />
+        <PackageReference Include="DynamoVisualProgramming.Core" Version="3.3.0.6316" ExcludeAssets="runtime" GeneratePathProperty="true" />
```
ah, I see your description now.

pinzart90 commented on src/Config/CS_SDK.props:
```diff
@@ -1,11 +1,10 @@
 <Project>
 	<Import Project="$(SolutionDir)Config/user_local.props" />
 	<PropertyGroup>
-		<Platforms>NET70;NET80</Platforms>
```
CICD can invoke a build either with platform values set, or with other underlying properties (UI vs Core)

pinzart90 commented on src/DADynamoApp/DADynamoApp.csproj:
```diff
@@ -0,0 +1,47 @@
+    </ImportGroup>
+    
+  <PropertyGroup>
+      <TargetFramework>net8.0-windows</TargetFramework>
```
yes, visual studio keeps putting it in here...

pinzart90 commented on src/DADynamoApp/DADynamoApp.csproj:
```diff
@@ -0,0 +1,47 @@
+        </Reference>
+    </ItemGroup>
+    <ItemGroup>
+        <PackageReference Include="Autodesk.Forge.DesignAutomation.Revit" Version="2025.0.1" />
```
seems like a 2026 version of this package was pushed two weeks ago. Not sure if the versions exactly match.

pinzart90 commented on src/DADynamoApp/DADynamoApp.csproj:
```diff
@@ -0,0 +1,47 @@
+      <ProjectReference Include="..\Libraries\RevitServices\RevitServices.csproj" />
+    </ItemGroup>
+    <ItemGroup>
+        <Reference Include="DynamoPlayer.Models">
```
because the player package does not respect the dotnet norms. It is built like a revit nuget (weird dll paths)

pinzart90 commented on src/DADynamoApp/DAEntrypoint.cs:
```diff
@@ -0,0 +1,182 @@
+namespace DADynamoApp
+{
+    [Autodesk.Revit.Attributes.Regeneration(Autodesk.Revit.Attributes.RegenerationOption.Manual)]
+    [Autodesk.Revit.Attributes.Transaction(Autodesk.Revit.Attributes.TransactionMode.Manual)]
```
I am not sure. @Mikhinja could you shed some light ? what would be better to use ?
Is automatic transaction mode auto triggered after every change ?

Mikhinja commented on src/Config/CS_SDK.props:
```diff
@@ -1,11 +1,10 @@
+		<Platforms>NET70;NET80;NET80_DA</Platforms>
 		<PlatformTarget >x64</PlatformTarget>
-		<TargetFramework Condition="'$(Platform)' == 'NET80'">net8.0-windows</TargetFramework>
-		<TargetFramework Condition="'$(Platform)' == 'NET70'">net7.0-windows</TargetFramework>
```
Confirm, removable.

pinzart90 committed:
Merge branch 'master' into D4R_DA

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
Update AssemblyInfoGenerator.csproj

pinzart90 committed:
Update AssemblyInfoGenerator.csproj

pinzart90 committed:
update

pinzart90 committed:
Update DynamoRevitIcons.csproj

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

add saveRvt option

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
Update DAEntrypoint.cs

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
Update DAEntrypoint.cs

pinzart90 committed:
update

pinzart90 committed:
Update RevitServices.csproj

pinzart90 committed:
Update RevitServices.csproj

pinzart90 committed:
Update DocumentManager.cs

pinzart90 committed:
Merge branch 'master' into D4R_DA

pinzart90 committed:
update for latest master

pinzart90 committed:
simplify

pinzart90 committed:
update

pinzart90 committed:
Update DynamoRevitApp.cs

pinzart90 committed:
Add cloud model support (#3252)

What changed:
Ability to load Revit Cloud Models
("type":"versions:autodesk.bim360:C4RModel")

Documentation 

https://aps.autodesk.com/blog/new-sample-support-access-revit-cloud-model-revit-design-automation-engine

Not implemented yet: Cloud Shared models.
Does not work for regular rvt files stored on ACC (only supported in
Autocad for now) -
https://aps.autodesk.com/en/docs/design-automation/v3/developers_guide/reference-downloading/

TODO:
To access cloud models using Automation, you need to provide the user
context through a special workitem argument; adsk3LeggedToken,
scope=code:all needs to be added to your authorization call.
I need to add an example in the scripts/postman collections

pinzart90 committed:
Update DAEntrypoint.cs

pinzart90 committed:
update

pinzart90 commented on src/DADynamoApp/DADynamoApp.csproj:
```diff
@@ -0,0 +1,47 @@
+  </PropertyGroup>
+    
+  <ItemGroup>
+    <Compile Remove="obj\**" />
```
old code from when I was still experimenting with this project, will remove

pinzart90 committed:
update

pinzart90 committed:
update

pinzart90 committed:
update for 2026

Temporary changes until we rebase to 2026 or use revit 2027

mjkkirschner committed:
WIP - unify log output between DA appbundle and daas | 2026 (#3296)

~~⚠️ do not merge while this branch targets net8 / dynamo 3.5 - can only
ref dynamo 4.0~~ Retargeted to work in Revit 2026 / Dynamo 3.5.

I will send another PR and create a new branch for D4R_DA_2027

untested!

* `service mode` already redirects logger to std out.
* we inject some profiling and send them to std out so they are captured
in da logs.
*  test
* added daapp.csproj to revit.all.sln - but unsure on this.
* added a TODO file because I think there is a lot of cleanup required
in this repo.



@pinzart @pinzart90 I am thinking this branch (`D4R_DA`) gets renamed to
D4R_DA_2026 and we rip out all the NET10 stuff - sound good?

pinzart90 committed:
Isolate the DA Application class (#3299)

ISolate the DA Application class from the rest of the implementation so
that we can handle setup an Assembly resolver before all the
dependencies need to get solved.


Please Note:
1. Before submitting the PR, please review [How to Contribute to
Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github
(Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR
has 30 days to respond. If the PR has seen no activity by the next
session, it will be either closed by the team or depending on its
utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these
other template
options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template)
in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled
out with all declarations satisfied will not be reviewed by the Dynamo
team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into
all the DynamoRevit Release branches that Dynamo supports. Contributors
will be responsible for cherry-picking their reviewed commits to the
other branches after a `LGTM` label is added to the PR.

### Purpose

(FILL ME IN) This section describes why this PR is here. Usually it
would include a reference
to the tracking task that it is part or all of the solution for.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the
[standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] Snapshot of UI changes, if any.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

pinzart90 committed:
update for transactions (#3303)

### Purpose

https://jira.autodesk.com/browse/DYN-10310
Cannot save rvt after running graph

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the
[standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] Snapshot of UI changes, if any.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of
